TheAmazingAudioEngine/AEAudioUnitFileStreamer.m

```diff
@@ -0,0 +1,395 @@
+    
+    // Setup play region
+    error = nil;
+    [self setupPlayRegion:&error];
+    if(error) {
+        NSLog(@"Error: %@",[error description]);
+    }
+    
+    // set our own audio controller property so we can use it during the postrender callback
+    _audioController = audioController;
+    
+    // Setup the postrender callback
+    AudioUnitAddRenderNotify(AEAudioUnitChannelGetAudioUnit(self), postRenderCallback, (__bridge void *)(self));
+}
+
+- (id) initWithURL:(NSURL *)url audioController:(AEAudioController *)audioController error:(NSError **)error
+{
+    // Verify that the URL is valid and set the url
+    if(![[NSFileManager defaultManager] fileExistsAtPath:[url path]]) {
+        *error = [NSError errorWithDomain: NSURLErrorDomain code:-50 userInfo:[NSDictionary dictionaryWithObject:@"URL is invalid" forKey:NSLocalizedDescriptionKey]];
+        return nil;
+    }
+    
+    AudioComponentDescription aedesc = AEAudioComponentDescriptionMake(kAudioUnitManufacturer_Apple, kAudioUnitType_Generator, kAudioUnitSubType_AudioFilePlayer);
+    if(self = [super initWithComponentDescription:aedesc]) {
+        
+        // we don't want to start playing immediately.
+        self.channelIsPlaying = NO;
+        
+        _url = url;
+    }
+    
+    return(self);
+}
+
+- (BOOL) setupURL: (NSURL*) url error:(NSError**) error{
+    // Load the URL
+    OSStatus result;
+    ExtAudioFileRef audioFileRef;
+    checkResult(result = ExtAudioFileOpenURL((CFURLRef)CFBridgingRetain(url), &audioFileRef), "AudioFileOpenURL");
+    if(result == noErr)
+    {
+        // Get the audio file itself
+        UInt32 size = sizeof(_audioUnitFile);
+        checkResultForError(ExtAudioFileGetProperty(audioFileRef, kExtAudioFileProperty_AudioFile, &size, &_audioUnitFile), "ExtAudioFileGetProperty(kExtAudioFileProperty_AudioFile)", "Failed getting audio file from extAudioFileRef", NO, result, error)
+        
+        // Set the file to play
+        AudioUnit au = AEAudioUnitChannelGetAudioUnit(self);
+        checkResultForError(AudioUnitSetProperty(au, kAudioUnitProperty_ScheduledFileIDs, kAudioUnitScope_Global, 0, &_audioUnitFile, sizeof(_audioUnitFile)), "AudioUnitSetProperty(kAudioUnitProperty_ScheduledFileIDs)", "Failed setting audio unit file IDs", NO, result, error)
+        
+        // Get audio description
+        size = sizeof(_audioDescription);
+        checkResultForError(AudioFileGetProperty(_audioUnitFile, kAudioFilePropertyDataFormat, &size, &_audioDescription), "AudioFileGetProperty(kAudioFilePropertyDataFormat)", "Failed getting audio file data format", NO, result, error)
+        
+        // Get file length
+        UInt64 fileLengthInFrames = 0;
+        size = sizeof(fileLengthInFrames);
+        checkResultForError(ExtAudioFileGetProperty(audioFileRef, kExtAudioFileProperty_FileLengthFrames, &size, &fileLengthInFrames), "ExtAudioFileGetProperty(kExtAudioFileProperty_AudioFile)", "Failed getting file length", NO, result, error)
```

Comment:
Not totally sure on this one, but I believe ExtAudioFile takes care of that. We tested with TheEngineSample's "Southern Rock Drums.m4a", and _lengthInFrames == 88200, which appears to match exactly with the audio data of that file when opened in a DAW. What's the proper way to check this?